### DISCUSSION TITLE
Doubt in Polygon Contains point

### DISCUSSION THREAD

**Discussion by sreekanth-cb** [google/s2geometry#263] [Q&A] [answered]
Hi @jmr ,

```
TEST(S2NonIntersecting, AllCases) {
  vector<S2Point> pts(9);
  S2LatLng latlng;

  pts[0] = latlng.FromDegrees(18.769449106345064, 70.51240102521727).ToPoint();
  pts[1] = latlng.FromDegrees(19.819449106345065, 72.01240102521727).ToPoint();
  pts[2] = latlng.FromDegrees(23.769449106345064, 72.01240102521727).ToPoint();
  pts[3] = latlng.FromDegrees(26.819449106345065, 72.01240102521727).ToPoint();
  pts[4] = latlng.FromDegrees(30.769449106345064, 72.01240102521727).ToPoint();
  pts[5] = latlng.FromDegrees(31.819449106345065, 72.01240102521727).ToPoint();
  pts[6] = latlng.FromDegrees(34.769449106345064, 72.01240102521727).ToPoint();
  pts[7] = latlng.FromDegrees(36.81944910634506, 72.01240102521727).ToPoint();
  pts[8] = latlng.FromDegrees(39.769449106345064, 72.01240102521727).ToPoint();
  
  S2Loop* loop = new S2Loop(pts);
  EXPECT_EQ(true, loop->IsValid());
  EXPECT_EQ(true, loop->IsNormalized());

  vector<std::unique_ptr<S2Loop>> loops;
  loops.emplace_back(loop);  
  S2Polygon polygon1 = S2Polygon(std::move(loops));

  EXPECT_EQ(true, polygon1.IsValid());
  
  // points lies outside
  S2Point point = latlng.FromDegrees(32.87762784238387, 71.45855236113707).ToPoint();
  //S2Point point = latlng.FromDegrees(32.39851580247402, 68.741455078125).ToPoint(); // returns right results.
  
  bool contains = polygon1.Contains(point);
  EXPECT_EQ(true, contains); 
}
```

Even though the points lie outside the polygon, it returns that the point is contained within. (only for 1 point that lies close to the edge)

These polygon coordinates are validated as legit vertices here https://geojsonlint.com/ 
Hence I don't know whether it's a CCW order issue.

Any clues on what is going wrong here? 

Thanks,

**Comment by jmr** [comment, marked as answer]
> Even though the points lie outside the polygon, 

How do you know the point is outside? `S2Polygon::Contains(S2Point)` is well exercised and unlikely to be wrong. It looks like your point is close to the boundary. `S1Angle(point, polygon1.Project(point) ~ 0`.

**Reply by sreekanth-cb** [reply]
Yes @jmr, it is indeed close to the boundary when I was projecting it using the [geojson](https://geojson.io/) tool. 
But I realize that the projection won't be taking care of the geodesic/curvature while projecting to the earth's sphere.
Perhaps that attributes to this difference in the online projection.

Do you have any clues about verifying the results (or any libs/tools) for such cases for actual containment relationships?

**Reply by jmr** [reply]
It looks like geojson.io doesn't draw edges using geodesics. Assume S2 is right.

**Reply by sreekanth-cb** [reply]
thank you @jmr.
